src/go/k8s/controllers/redpanda/cluster_controller.go

```diff
@@ -343,86 +341,112 @@ func (r *ClusterReconciler) WithClusterDomain(
+			externalAdminListener != nil && needExternalIP(externalAdminListener.External) ||
+			externalProxyListener != nil && needExternalIP(externalProxyListener.External) ||
+			schemaRegistryConf != nil && schemaRegistryConf.External != nil && needExternalIP(*schemaRegistryConf.External) {
 			if err := r.Get(ctx, types.NamespacedName{Name: pods[i].Spec.NodeName}, &node); err != nil {
-				return []string{}, []string{}, []string{}, "", fmt.Errorf("failed to retrieve node %s: %w", pods[i].Spec.NodeName, err)
+				return nil, fmt.Errorf("failed to retrieve node %s: %w", pods[i].Spec.NodeName, err)
 			}
 		}
 
 		if externalKafkaListener != nil && len(externalKafkaListener.External.Subdomain) > 0 {
 			address := subdomainAddress(podName, externalKafkaListener.External.Subdomain, getNodePort(&nodePortSvc, resources.ExternalListenerName))
-			observedNodesExternal = append(observedNodesExternal, address)
+			result.External = append(result.External, address)
 		} else if externalKafkaListener != nil {
-			observedNodesExternal = append(observedNodesExternal,
+			result.External = append(result.External,
 				fmt.Sprintf("%s:%d",
 					getExternalIP(&node),
 					getNodePort(&nodePortSvc, resources.ExternalListenerName),
 				))
 		}
-		if externalAdminListener != nil && len(externalKafkaListener.External.Subdomain) > 0 {
+
+		if externalAdminListener != nil && len(externalAdminListener.External.Subdomain) > 0 {
 			address := subdomainAddress(podName, externalAdminListener.External.Subdomain, getNodePort(&nodePortSvc, resources.AdminPortExternalName))
-			observedNodesExternalAdmin = append(observedNodesExternalAdmin, address)
+			result.ExternalAdmin = append(result.ExternalAdmin, address)
 		} else if externalAdminListener != nil {
-			observedNodesExternalAdmin = append(observedNodesExternalAdmin,
+			result.ExternalAdmin = append(result.ExternalAdmin,
 				fmt.Sprintf("%s:%d",
 					getExternalIP(&node),
 					getNodePort(&nodePortSvc, resources.AdminPortExternalName),
 				))
 		}
-		if externalProxyListener != nil && len(externalKafkaListener.External.Subdomain) > 0 {
+
+		if externalProxyListener != nil && len(externalProxyListener.External.Subdomain) > 0 {
 			address := subdomainAddress(podName, externalProxyListener.External.Subdomain, getNodePort(&nodePortSvc, resources.PandaproxyPortExternalName))
-			observedNodesExternalProxy = append(observedNodesExternalProxy, address)
+			result.ExternalPandaproxy = append(result.ExternalPandaproxy, address)
 		} else if externalProxyListener != nil {
-			observedNodesExternalProxy = append(observedNodesExternalProxy,
+			result.ExternalPandaproxy = append(result.ExternalPandaproxy,
 				fmt.Sprintf("%s:%d",
 					getExternalIP(&node),
 					getNodePort(&nodePortSvc, resources.PandaproxyPortExternalName),
 				))
 		}
+
+		if schemaRegistryConf != nil && schemaRegistryConf.External != nil && schemaRegistryConf.External.Subdomain == "" {
```

Comment:
why is needExternalIP not used here?

Reply:
Because it's not used in any of the previous checks.

I will change it to all of them